flink-ml-lib/src/main/java/org/apache/flink/ml/feature/featurehasher/FeatureHasher.java

```diff
@@ -0,0 +1,143 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+import static org.apache.flink.shaded.guava30.com.google.common.hash.Hashing.murmur3_32;
+
+/**
+ * FeatureHasher projects a set of categorical or numerical features into a vector with fixed
+ * dimension. This is done using the hashing trick (https://en.wikipedia.org/wiki/Feature_hashing)
+ * to map features to indices in the feature vector.
+ */
+public class FeatureHasher
+        implements Transformer<FeatureHasher>, FeatureHasherParams<FeatureHasher> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+    private static final HashFunction HASH = murmur3_32(0);
+
+    public FeatureHasher() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new HashFunc(
+                                        getInputCols(), getCategoricalCols(), getNumFeatures()),
+                                outputTypeInfo);
+        Table outputTable = tEnv.fromDataStream(output);
+        return new Table[] {outputTable};
+    }
+
+    private static class HashFunc implements MapFunction<Row, Row> {
+        private final String[] categoricalCols;
+        private final int numFeatures;
+        private final String[] numericCols;
+
+        public HashFunc(String[] inputCols, String[] handleInvalid, int numFeatures) {
+            this.categoricalCols = handleInvalid;
+            this.numFeatures = numFeatures;
+            this.numericCols = ArrayUtils.removeElements(inputCols, categoricalCols);
+        }
+
+        @Override
+        public Row map(Row row) {
+            TreeMap<Integer, Double> feature = new TreeMap<>();
+            for (String col : numericCols) {
+                if (null != row.getField(col)) {
+                    double value = ((Number) row.getFieldAs(col)).doubleValue();
+                    updateMap(col, value, feature, numFeatures);
+                }
+            }
+            for (String col : categoricalCols) {
+                if (null != row.getField(col)) {
+                    updateMap(col + "=" + row.getField(col), 1.0, feature, numFeatures);
+                }
+            }
+            return Row.join(row, Row.of(new SparseVector(numFeatures, feature)));
```

Comment:
It might be better to extract the indices and values from the TreeMap here and use the original constructor of `SparseVector`. I'm not sure the newly-introduced SparseVector that accepts a TreeMap would be widely used.

Reply:
In Alink, this constructor with Map is used by ItemCF, Naive Bayes, FeatureHasher and vectorAssembler. 

SparseVector also called kv data structure. Then I add this constructor.

Reply:
Got it. Then I'm OK with adding this constructor.

Reply:
I hold a bit different opinion. It is kind of wierd to me to let SparseVector accept a `map` as a parameter for the constructor. 

Essentially what we need is a utility function that transforms a Map<Integer, Double> to `int[] indices, double[] values`.  What do you think @yunfengzhou-hub @weibozhao ?

Reply:
I just noticed that we have already implemented NaiveBayes and VectorAssembler without this constructor. If the new constructor is only used in FeatureHasher and ItemCF, I would prefer not to add this constructor.

Reply:
I agree with you @zhipeng93, I will just add the transformation code in my FeatureHasher algorithm, if some other algorithm use this transform, then make it a common function.